pkg/ruler/rulestore/bucket_client.go

```diff
@@ -0,0 +1,312 @@
+package rulestore
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	"github.com/cortexproject/cortex/pkg/tenant"
+)
+
+const (
+	delim      = "/"
+	rulePrefix = "rules"
+)
+
+type BucketRuleStore struct {
+	bucket          objstore.Bucket
+	loadConcurrency int
+
+	logger log.Logger
+}
+
+func NewBucketRuleStore(bucket objstore.Bucket, loadConcurrency int, logger log.Logger) *BucketRuleStore {
+	return &BucketRuleStore{
+		bucket:          bucket,
+		loadConcurrency: loadConcurrency,
+		logger:          logger,
+	}
+}
+
+func newRuleStoreUserBucketClient(userID string, bkt objstore.Bucket) objstore.Bucket {
+	return bucket.NewUserBucketClient(userID, bucket.NewPrefixedBucketClient(bkt, rulePrefix))
+}
+
+func (b *BucketRuleStore) listNamespacesForUser(ctx context.Context, user string) ([]string, error) {
+	userBucket := newRuleStoreUserBucketClient(user, b.bucket)
+
+	var namespaces []string
+	err := userBucket.Iter(ctx, "", func(namespace string) error {
+		namespace = strings.TrimSuffix(namespace, delim)
+		decodedNamespace, err := base64.URLEncoding.DecodeString(namespace)
+		if err != nil {
+			return err
+		}
+
+		namespaces = append(namespaces, string(decodedNamespace))
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return namespaces, nil
+}
+
+// If existing rule group is supplied, it is Reset and reused. If nil, new RuleGroupDesc is allocated.
+func (b *BucketRuleStore) getRuleGroup(ctx context.Context, userID, namespace, groupName string, rg *rules.RuleGroupDesc) (*rules.RuleGroupDesc, error) {
+	userBucket := newRuleStoreUserBucketClient(userID, b.bucket)
+	objectKey := generateRuleObjectKey(namespace, groupName)
+
+	reader, err := userBucket.Get(ctx, objectKey)
+	if userBucket.IsObjNotFoundErr(err) {
+		level.Debug(b.logger).Log("msg", "rule group does not exist", "name", objectKey)
+		return nil, rules.ErrGroupNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = reader.Close() }()
+
+	buf, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	if rg == nil {
+		rg = &rules.RuleGroupDesc{}
+	} else {
+		rg.Reset()
+	}
+
+	err = proto.Unmarshal(buf, rg)
+	if err != nil {
+		return nil, err
+	}
+
+	return rg, nil
+}
+
+func (b *BucketRuleStore) ListAllUsers(ctx context.Context) ([]string, error) {
+	var users []string
+	err := b.bucket.Iter(ctx, rulePrefix, func(user string) error {
+		user = strings.TrimSuffix(strings.TrimPrefix(user, rulePrefix+delim), delim)
+		iterErr := tenant.ValidTenantID(user)
```

Comment:
We don't do this validation in objectclient RuleStore. Does it really belong here?

Reply:
👍  Makes sense, the context on tenants should be left to high level modules.